opm/autodiff/AquiferFetkovich.hpp

```diff
@@ -21,367 +21,214 @@ along with OPM.  If not, see <http://www.gnu.org/licenses/>.
-      const std::unordered_map<int, int>& cartesian_to_compressed,
-      const Simulator& ebosSimulator)
-      : ebos_simulator_ (ebosSimulator)
-      , aqufetp_data_ (aqufetp_data)
-      , cartesian_to_compressed_(cartesian_to_compressed)
-      , connection_ (connection)
-      {}
-
-        void initialSolutionApplied()
-        {
-          initQuantities(connection_);
-        }
-
-        void beginTimeStep()
-        {
-          ElementContext elemCtx(ebos_simulator_);
-          auto elemIt = ebos_simulator_.gridView().template begin<0>();
-          const auto& elemEndIt = ebos_simulator_.gridView().template end<0>();
-          for (; elemIt != elemEndIt; ++elemIt) {
-            const auto& elem = *elemIt;
-
-            elemCtx.updatePrimaryStencil(elem);
-
-            int cellIdx = elemCtx.globalSpaceIndex(0, 0);
-            int idx = cellToConnectionIdx_[cellIdx];
-            if (idx < 0)
-            continue;
-
-            elemCtx.updateIntensiveQuantities(0);
-            const auto& iq = elemCtx.intensiveQuantities(0, 0);
-            pressure_previous_[idx] = Opm::getValue(iq.fluidState().pressure(waterPhaseIdx));
-          }
-        }
-
-        template <class Context>
-        void addToSource(RateVector& rates, const Context& context, unsigned spaceIdx, unsigned timeIdx)
-        {
-          unsigned cellIdx = context.globalSpaceIndex(spaceIdx, timeIdx);
-
-          int idx = cellToConnectionIdx_[cellIdx];
-          if (idx < 0)
-          return;
-
-          // We are dereferencing the value of IntensiveQuantities because cachedIntensiveQuantities return a const pointer to
-          // IntensiveQuantities of that particular cell_id
-          const IntensiveQuantities intQuants = context.intensiveQuantities(spaceIdx, timeIdx);
-          // This is the pressure at td + dt
-          updateCellPressure(pressure_current_,idx,intQuants);
-          updateCellDensity(idx,intQuants);
-          calculateInflowRate(idx, context.simulator());
-          rates[BlackoilIndices::conti0EqIdx + FluidSystem::waterCompIdx] +=
-          Qai_[idx]/context.dofVolume(spaceIdx, timeIdx);
-        }
+                      const Aquancon::AquanconOutput& connection,
+                      const std::unordered_map<int, int>& cartesian_to_compressed,
+                      const Simulator& ebosSimulator)
+    : aqufetp_data_(aqufetp_data)
+    , Base(connection, cartesian_to_compressed, ebosSimulator)
+    {}
+
+    void endTimeStep()
+    {
+      for (const auto& Qai: Base::Qai_) {
+        Base::W_flux_ += Qai*Base::ebos_simulator_.timeStepSize();
+        aquifer_pressure_ = aquiferPressure();
+      }
+    }
+
+      protected:
+        // Aquifer Fetkovich Specific Variables
+        const Aquifetp::AQUFETP_data aqufetp_data_;
+        Scalar aquifer_pressure_; // aquifer
 
-        void endTimeStep()
+        inline void initializeConnections(const Aquancon::AquanconOutput& connection)

```

Comment:
Please also check whether this function can go to the base class.

Reply:
This cannot as it uses the input deck data. Therefore it is specific to the object. Maybe, this can changed in the future but for now I will leave it like this.